From f1a55b2f0b22a152d430dce53b2b50e4e5c5bca2 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Tue, 23 Jul 2019 16:14:00 +0200
Subject: [PATCH 356/386] telephony: mvebu_phone: enable tal_dev via module
 parameter

Exposing the TDM controller via tal_dev miscdevice is now
optional. By default it is disabled.

In order to enable it, pass "mv_phone.enable_tal_dev=1" via
command-line. As a result /dev/talX, where X is TDM controller's
index, will be created, which can be used for user-space telephony
applications.

Update the documentation about tal_dev example of using
the multiple TDM controllers.

Change-Id: I3ddfbe09dc2428978e6b2d2325c68571896f4012
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/12964
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone_dev.c  | 14 ++++++++++----
 .../mvebu_phone/telephony_user_guide.txt      | 19 +++++++++++++++++++
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index 7f493f648951..4bd0319dcf56 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -18,6 +18,12 @@
 
 #define DRV_NAME "mvebu_phone"
 
+/* Module parameters */
+static bool enable_tal_dev;
+module_param(enable_tal_dev, bool, 0444);
+MODULE_PARM_DESC(enable_tal_dev,
+		 "Expose TDM via char device: 0:disable(default), 1:enable");
+
 /* Statistic printout in userspace via /proc/tdm */
 static int mv_phone_status_show(struct seq_file *m, void *v)
 {
@@ -969,7 +975,7 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 		priv->irq_count = 1;
 
 		err = tal_set_if(&tdm2c_if, &pdev->dev, priv->id,
-				 &priv->miscdev, true);
+				 &priv->miscdev, enable_tal_dev);
 		if (err)
 			goto err_axi_clk;
 	}
@@ -984,7 +990,7 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 		priv->tdmmc_ip_ver = TDMMC_REV1;
 
 		err = tal_set_if(&tdmmc_if, &pdev->dev, priv->id,
-				 &priv->miscdev, true);
+				 &priv->miscdev, enable_tal_dev);
 		if (err)
 			goto err_axi_clk;
 	}
@@ -997,7 +1003,7 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 		priv->tdmmc_ip_ver = TDMMC_REV1;
 
 		err = tal_set_if(&tdmmc_if, &pdev->dev, priv->id,
-				 &priv->miscdev, true);
+				 &priv->miscdev, enable_tal_dev);
 		if (err)
 			goto err_axi_clk;
 	}
@@ -1063,7 +1069,7 @@ static int mvebu_phone_remove(struct platform_device *pdev)
 {
 	struct mv_phone_dev *priv = dev_get_drvdata(&pdev->dev);
 
-	tal_set_if(NULL, priv->dev, priv->id, &priv->miscdev, true);
+	tal_set_if(NULL, priv->dev, priv->id, &priv->miscdev, enable_tal_dev);
 
 	clk_disable_unprepare(priv->clk);
 	clk_disable_unprepare(priv->axi_clk);
diff --git a/drivers/telephony/mvebu_phone/telephony_user_guide.txt b/drivers/telephony/mvebu_phone/telephony_user_guide.txt
index 1ba7135491ee..8c0da3c8bfcc 100644
--- a/drivers/telephony/mvebu_phone/telephony_user_guide.txt
+++ b/drivers/telephony/mvebu_phone/telephony_user_guide.txt
@@ -50,6 +50,25 @@ For example::
 	> insmod silabs-si3226x.ko
 
 
+Accessing multiple TDM controllers
+----------------------------------
+In order to make use of the user-space telephony applications, it is
+possible to expose the TDM interfaces via /dev/talX character devices
+(X stands for the TDM controller id, as specified by 'cell-index'
+DT property).
+
+Above option is disabled by default. In order to enable it, pass
+mv_phone.enable_tal_dev=1" via U-Boot command line when booting
+the Linux kernel.
+
+The example uses already implemented driver:
+drivers/telephony/mvebu_phone/tal/tal_dev.c
+However, depending on the architecture and requirements, the TAL
+(Telephony Adaptation Layer) callback routines:
+  - tal_set_if()
+  - tal_dev_init/exit()
+allow to register and connect the TDM controllers with other drivers.
+
 mv-voice-tool
 -------------
 User-space application which communicates via ioctl with the telephony driver.
-- 
2.17.1

